### PR TITLE
Fix false positive cycle detection

### PR DESCRIPTION
**Pull request by kwridan**
Resolves https://github.com/tuist/tuist/issues/460

### Short description 📝

In some cases false cycle errors were being thrown during project generation:

- A -> B
- C -> A
- C -> B

### Solution 📦

- Implemented depth first search to check for cycles once the graph is fully constructed
- Changed the order in which target nodes are added to the cache to prevent infinite loops in the event of cycles being present

### Implementation 👩‍💻👨‍💻

- [x] Update `GraphCircularDetecting`
- [x] Update `GraphCircularDetectorTests`
- [x] Update graph loading error to cache nodes ahead of traversing their dependencies (to avoid infinite loops!)
- [x] Add fixture
- [x] Update change log

### Test Plan 🛠

- Run `tuist generate` within `fixtures/ios_workspace_with_dependency_cycle`
- Verify a cycle error is displayed
- Verify generating other fixtures still works as before

### CODE REVIEW

File: Sources/TuistCoreTesting/Extensions/XCTestCase+Extras.swift [hunk 1]
```diff
@@ -59,9 +59,12 @@ public extension XCTestCase {
             _ = try closure()
         } catch let closureError as Error {
             XCTAssertEqual(error, closureError, file: file, line: line)
+            return
```
Comment (pepibumur) [inline Sources/TuistCoreTesting/Extensions/XCTestCase+Extras.swift]: Good catch here

File: Sources/TuistGenerator/Graph/GraphLoadingError.swift [hunk 2]
```diff
@@ -40,12 +40,9 @@ enum GraphLoadingError: FatalError, Equatable {
...
             return message
-        case let .circularDependency(from, to):
-            var message = ""
-            message.append("Found circular dependency between the target")
-            message.append(" '\(from.name)' at '\(from.path.pathString)'")
-            message.append(" and the target '\(to.name)' at '\(to.path.pathString)'")
-            return message
+        case let .circularDependency(nodes):
+            let nodeDescriptions = nodes.map { "\($0.path):\($0.name)" }
+            return "Found circular dependency between targets: \(nodeDescriptions.joined(separator: " -> "))"
```
Comment (pepibumur) [inline Sources/TuistGenerator/Graph/GraphLoadingError.swift]: 👏

File: Tests/TuistGeneratorTests/Graph/GraphCircularDetectorTests.swift [hunk 3]
```diff
@@ -4,25 +4,190 @@ import XCTest
...
+        subject.start(from: a, to: b)
+        subject.start(from: a, to: c)
+        subject.start(from: b, to: c)
+
+        subject.start(from: d, to: e)
+        subject.start(from: e, to: f)
+        subject.start(from: f, to: d)
+
+        // Then
+        XCTAssertThrowsError(try subject.complete())
```
Comment (pepibumur) [inline Tests/TuistGeneratorTests/Graph/GraphCircularDetectorTests.swift]: I'd assert that we are throwing the right error.

Comment (kwridan) [reply Tests/TuistGeneratorTests/Graph/GraphCircularDetectorTests.swift]: 👍 indeed otherwise it could be flagging the wrong nodes! will update it

**Comment by pepibumur** [comment]
Ignore the Netlify errors @kwridan

**Comment by pepibumur** [comment]
Thanks for tackling this @kwridan

**Comment by kwridan** [comment]
👍 was a fun one :)

**Pull request merged**